File: core/src/main/java/javabot/dao/ChannelDao.java
package javabot.dao;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

import org.mongodb.morphia.query.Query;
import org.mongodb.morphia.query.QueryImpl;
import javabot.model.Activity;
import javabot.dao.util.QueryParam;
import javabot.model.Channel;
import javabot.model.criteria.ChannelCriteria;
import org.apache.commons.lang.StringUtils;
import org.bson.types.ObjectId;

@SuppressWarnings({"ConstantNamingConvention"})
public class ChannelDao extends BaseDao<Channel> {
  public ChannelDao() {
    super(Channel.class);
  }

  public void delete(final ObjectId id) {
    delete(find(id));
  }

  @SuppressWarnings({"unchecked"})
  public List<String> configuredChannels() {
    return ((QueryImpl) getQuery()).getCollection().distinct("name");
  }

  @SuppressWarnings({"unchecked"})
  public List<Channel> getChannels() {
    return getChannels(false);
  }

  public List<Channel> getChannels(Boolean showAll) {
    ChannelCriteria channelCriteria = new ChannelCriteria(ds);
    if (!showAll) {
      channelCriteria.logged().equal(true);
    }
    Query<Channel> query = channelCriteria.query();
    query.order("name");
    return query.asList();
  }

  @SuppressWarnings({"unchecked"})
  public List<Channel> find(final QueryParam qp) {
    String condition = qp.getSort();
    if (!qp.isSortAsc()) {
      condition = "-" + condition;
    }
    return getQuery().order(condition).asList();
  }

  public boolean isLogged(final String channel) {
    final Channel chan = get(channel);
    return chan != null ? chan.getLogged() : Boolean.FALSE;
  }

  public Channel get(final String name) {
    ChannelCriteria criteria = new ChannelCriteria(ds);
    criteria.upperName().equal(name.toUpperCase());
    return criteria.query().get();
  }

  @SuppressWarnings({"unchecked"})
  public List<Activity> getStatistics() {
/*
@NamedQuery(name = ChannelDao.STATISTICS, query = "select new javabot.model.Activity(l.channel, count(l), max(l.updated),"
    + " min(l.updated), (select count(e) from Logs e)) from Logs l "
    + "where l.channel like '#%' group by l.channel order by count(l) desc")
*/
    return null;//(List<Activity>) getEntityManager().createNamedQuery(ChannelDao.STATISTICS)
//            .getResultList();
  }

  @SuppressWarnings({"unchecked"})
  public List<String> loggedChannels() {
    ChannelCriteria criteria = new ChannelCriteria(ds);
    criteria.logged().equal(true);
    List<Channel> channels = criteria.query().retrievedFields(true, "name").asList();
    List<String> names = new ArrayList<>();
    for (Channel channel : channels) {
      names.add(channel.getName());
    }
    return names;
  }

  public Channel create(final String name, final Boolean logged, final String key) {
    final Channel channel = new Channel();
    channel.setName(name);
    channel.setLogged(logged == null ? Boolean.TRUE : logged);
    channel.setKey(!StringUtils.isEmpty(key) ? key : null);
    save(channel);
    return channel;
  }

  public void save(final Channel channel) {
    channel.setUpdated(LocalDateTime.now());
    super.save(channel);
  }

}

File: src/main/kotlin/javabot/dao/ChannelDao.kt
package javabot.dao

import com.google.inject.Inject
import dev.morphia.Datastore
import dev.morphia.query.FindOptions
import dev.morphia.query.Sort
import javabot.dao.util.QueryParam
import javabot.model.Activity
import javabot.model.Channel
import javabot.model.criteria.ChannelCriteria.Companion.logged
import javabot.model.criteria.ChannelCriteria.Companion.name
import javabot.model.criteria.ChannelCriteria.Companion.upperName
import org.apache.commons.lang.StringUtils
import java.time.LocalDateTime
import java.util.ArrayList
import java.util.Locale

@SuppressWarnings("ConstantNamingConvention")
class ChannelDao @Inject constructor(ds: Datastore) : BaseDao<Channel>(ds, Channel::class.java) {

    fun delete(name: String) {
        ds.find(Channel::class.java)
                .filter(name().eq(name))
                .delete()
    }

    @Suppress("UNCHECKED_CAST")
    fun configuredChannels(): List<String> {
        return ds.find(Channel::class.java)
                .iterator()
                .toList()
                .map { it.name }
    }

    fun getChannels(): List<Channel> {
        return getChannels(false)
    }

    fun getChannels(showAll: Boolean): List<Channel> {
        val query = ds.find(Channel::class.java)
        if (!showAll) {
            query.filter(logged().eq(true))
        }
        return query.iterator(FindOptions()
                        .sort(Sort.ascending(name)))
                .toList()
    }

    fun find(qp: QueryParam): List<Channel> {
        return getQuery().iterator(FindOptions()
                        .sort(qp.toSort()))
                .toList()
    }

    fun isLogged(channel: String): Boolean {
        val chan = get(channel)
        return chan?.logged ?: java.lang.Boolean.FALSE
    }

    fun get(name: String): Channel? {
        return ds.find(Channel::class.java)
                .filter(upperName().eq(name.uppercase(Locale.getDefault())))
                .first()
    }

    fun getStatistics(): List<Activity> {
        //        val criteria = ActivityCriteria(ds)
        /*
@NamedQuery(name = ChannelDao.STATISTICS, query = "select new javabot.model.Activity(l.channel, count(l), max(l.updated),"
    + " min(l.updated), (select count(e) from Logs e)) from Logs l "
    + "where l.channel like '#%' group by l.channel order by count(l) desc")
*/
        return listOf()//(List<Activity>) getEntityManager().createNamedQuery(ChannelDao.STATISTICS)
        //            .getResultList();
    }

    fun loggedChannels(): List<String> {
        val channels = ds.find(Channel::class.java)
                .filter(logged().eq(true)).iterator(FindOptions()
                        .projection().include(name))
                .toList()
        val names = ArrayList<String>()
        for (channel in channels) {
            names.add(channel.name)
        }
        return names
    }

    fun create(name: String, logged: Boolean?, key: String?): Channel {
        val channel = Channel()
        channel.name = name
        channel.logged = logged ?: java.lang.Boolean.TRUE
        channel.key = if (!StringUtils.isEmpty(key)) key else null
        save(channel)
        return channel
    }

    fun save(channel: Channel) {
        channel.updated = LocalDateTime.now()
        super.save(channel)
    }

    fun location(channel: Channel?): String {
        val location: String
        if (channel != null) {
            location = if (isLogged(channel.name)) channel.name else "private channel"
        } else {
            location = channel?.name ?: "private message"
        }
        return location
    }
}
